source/adapters/opencl/usm.cpp

```diff
@@ -207,20 +206,82 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueUSMFill(
-  ur_result_t RetVal = cl_ext::getExtFuncFromContext<clEnqueueMemFillINTEL_fn>(
-      CLContext, cl_ext::ExtFuncPtrCache->clEnqueueMemFillINTELCache,
-      cl_ext::EnqueueMemFillName, &FuncPtr);
+  if (patternSize <= 128) {
+    clEnqueueMemFillINTEL_fn EnqueueMemFill = nullptr;
+    UR_RETURN_ON_FAILURE(
+        cl_ext::getExtFuncFromContext<clEnqueueMemFillINTEL_fn>(
+            CLContext, cl_ext::ExtFuncPtrCache->clEnqueueMemFillINTELCache,
+            cl_ext::EnqueueMemFillName, &EnqueueMemFill));
+
+    CL_RETURN_ON_FAILURE(
+        EnqueueMemFill(cl_adapter::cast<cl_command_queue>(hQueue), ptr,
+                       pPattern, patternSize, size, numEventsInWaitList,
+                       cl_adapter::cast<const cl_event *>(phEventWaitList),
+                       cl_adapter::cast<cl_event *>(phEvent)));
+    return UR_RESULT_SUCCESS;
+  }
 
-  if (FuncPtr) {
-    RetVal = mapCLErrorToUR(
-        FuncPtr(cl_adapter::cast<cl_command_queue>(hQueue), ptr, pPattern,
-                patternSize, size, numEventsInWaitList,
-                cl_adapter::cast<const cl_event *>(phEventWaitList),
-                cl_adapter::cast<cl_event *>(phEvent)));
+  // OpenCL only supports pattern sizes as large as the largest CL type
+  // (double16/long16 - 128 bytes), anything larger we need to do on the host
+  // side and copy it into the target allocation.
+  clHostMemAllocINTEL_fn HostMemAlloc = nullptr;
+  UR_RETURN_ON_FAILURE(cl_ext::getExtFuncFromContext<clHostMemAllocINTEL_fn>(
+      CLContext, cl_ext::ExtFuncPtrCache->clHostMemAllocINTELCache,
+      cl_ext::HostMemAllocName, &HostMemAlloc));
+
+  clEnqueueMemcpyINTEL_fn USMMemcpy = nullptr;
+  UR_RETURN_ON_FAILURE(cl_ext::getExtFuncFromContext<clEnqueueMemcpyINTEL_fn>(
+      CLContext, cl_ext::ExtFuncPtrCache->clEnqueueMemcpyINTELCache,
+      cl_ext::EnqueueMemcpyName, &USMMemcpy));
+
+  clMemBlockingFreeINTEL_fn USMFree = nullptr;
+  UR_RETURN_ON_FAILURE(cl_ext::getExtFuncFromContext<clMemBlockingFreeINTEL_fn>(
+      CLContext, cl_ext::ExtFuncPtrCache->clMemBlockingFreeINTELCache,
+      cl_ext::MemBlockingFreeName, &USMFree));
+
+  cl_int ClErr = CL_SUCCESS;
+  auto HostBuffer = static_cast<uint64_t *>(
+      HostMemAlloc(CLContext, nullptr, size, 0, &ClErr));
+  CL_RETURN_ON_FAILURE(ClErr);
+
+  auto NumValues = size / sizeof(uint64_t);
+  auto NumChunks = patternSize / sizeof(uint64_t);
+  for (size_t i = 0; i < NumValues; i++) {
+    HostBuffer[i] = static_cast<const uint64_t *>(pPattern)[i % NumChunks];
   }
 
-  return RetVal;
+  cl_event CopyEvent = nullptr;
+  CL_RETURN_ON_FAILURE(USMMemcpy(
+      cl_adapter::cast<cl_command_queue>(hQueue), false, ptr, HostBuffer, size,
+      numEventsInWaitList, cl_adapter::cast<const cl_event *>(phEventWaitList),
+      &CopyEvent));
+
+  struct DeleteCallbackInfo {
+    clMemBlockingFreeINTEL_fn USMFree;
+    cl_context CLContext;
+    void *HostBuffer;
+    void execute() {
+      USMFree(CLContext, HostBuffer);
```

Comment:
Since I was adding a destructor anyway I gave the struct a proper constructor with does a retain on the context (with matching Release in the destructor), should prevent the released context scenario. Using a `new` pointer would probably work but it seems sketchier.